qiskit/qobj/qobj.py

```diff
@@ -26,42 +27,78 @@
 """
 
 
-class QobjSchema(BaseSchema):
-    """Schema for Qobj."""
+class QASMQobjSchema(BaseSchema):
+    """Schema for QASMQobj."""
 
     # Required properties.
     qobj_id = String(required=True)
-    config = Nested(QobjConfigSchema, required=True)
-    experiments = Nested(QobjExperimentSchema, required=True, many=True)
-    header = Nested(QobjHeaderSchema, required=True)
-    type = String(required=True,
-                  validate=OneOf(QobjType.QASM.value,
-                                 QobjType.PULSE.value))
+    config = Nested(QASMQobjConfigSchema, required=True)
+    experiments = Nested(QASMQobjExperimentSchema, required=True, many=True)
+    header = Nested(QASMQobjHeaderSchema, required=True)
+    type = String(validate=Equal(QobjType.QASM))
     schema_version = String(required=True, missing=QOBJ_VERSION)
 
 
-@bind_schema(QobjSchema)
-class Qobj(BaseModel):
-    """Model for Qobj.
+class PulseQobjSchema(BaseSchema):
+    """Schema for PulseQobj."""
+
+    # Required properties.
+    qobj_id = String(required=True)
+    config = Nested(PulseQobjConfigSchema, required=True)
+    experiments = Nested(PulseQobjExperimentSchema, required=True, many=True)
+    header = Nested(PulseQobjHeaderSchema, required=True)
+    type = String(validate=Equal(QobjType.PULSE))
+    schema_version = String(required=True, missing=QOBJ_VERSION)
+
+
+@bind_schema(QASMQobjSchema)
+class QASMQobj(BaseModel):
+    """Model for QASMQobj.
 
     Please note that this class only describes the required fields. For the
     full description of the model, please check ``QobjSchema``.
 
     Attributes:
         qobj_id (str): Qobj identifier.
-        config (QobjConfig): config settings for the Qobj.
-        experiments (list[QobjExperiment]): list of experiments.
-        header (QobjHeader): headers.
-        type (str): experiment type (QASM/PULSE).
+        config (QASMQobjConfig): config settings for the Qobj.
+        experiments (list[QASMQobjExperiment]): list of experiments.
+        header (QASMQobjHeader): headers.
+        schema_version (str): Qobj version.
+    """
+    def __init__(self, qobj_id, config, experiments, header, **kwargs):
+        # pylint: disable=redefined-builtin
+        self.qobj_id = qobj_id
+        self.config = config
+        self.experiments = experiments
+        self.header = header
+        self.type = 'QASM'
+
+        self.schema_version = QOBJ_VERSION
+
+        super().__init__(**kwargs)
+
+
+@bind_schema(PulseQobjSchema)
+class PulseQobj(BaseModel):
```

Comment:
Similarly `PulseQobj` and `QasmQobj` as well as all components that differ for `QASM` and `Pulse` like `PulseQobjConfig` and `QasmQobjConfig` so that type checking can still be useful. Ie. check if a given attribute is a general `Qobj`.